velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <deque>
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <TypeKind kind>
+void applyTypedFirstMatch(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+
+    auto search = searchDecoded.valueAt<T>(row);
```

Comment:
This wiki page says "By default, the function can assume that all inputs are not null for all “rows".": https://www.internalfb.com/intern/wiki/DataInfra/Velox/HowToWriteScalarFunction/#input-rows. So if I understand correctly, we don't need to worry about `search` being null because it is an input.

Reply:
Its says all inputs are not null , so some may be null ? If so couldnt search be null at row, but the array be non null ? 
Maybe it meant to say none of the inputs will be null ?

Reply:
Indeed. The document also says "By default, a vector function is assumed to have the default null behavior, e.g. null in any input produces a null result. In this case, the expression evaluation engine will exclude rows with nulls from the “rows” specified in the call to “apply”.". So I think it means none of the values at "rows" in the input vectors can be null when `apply` is being called.